ocaml/xapi/xapi_clustering.ml

```diff
@@ -132,19 +132,30 @@ let get_master_pif ~__context =
   | Some self -> Db.Cluster_host.get_PIF ~__context ~self
   | None -> raise Api_errors.(Server_error (internal_error, [ "No cluster_host exists on master" ]))
 
+(** Best-effort attempt to find a network common to the entire cluster *)
 let get_network_internal ~__context ~self =
-  let cluster_network = Db.PIF.get_network ~__context ~self:(get_master_pif ~__context) in
-  if List.exists
-    (fun cluster_host ->
-      let pif = Db.Cluster_host.get_PIF ~__context ~self:cluster_host in
-      let cluster_host_network = Db.PIF.get_network ~__context ~self:pif in
-      cluster_host_network <> cluster_network
-    ) (Db.Cluster_host.get_all ~__context)
-  then
-    debug "Not all cluster hosts of cluster %s on same network %s"
-      (Ref.string_of self) (Ref.string_of cluster_network);
-
-  cluster_network
+  let get_cluster_host_network self =
+    Db.Cluster_host.get_PIF ~__context ~self
+    |> (fun self -> Db.PIF.get_network ~__context ~self)
+  in
+  let common network = List.for_all (fun self -> network = (get_cluster_host_network self)) in
+  begin match find_cluster_host ~__context ~host:(Helpers.get_master ~__context) with
+  | Some cluster_host ->
+    info "cluster_host on master!";
+    get_cluster_host_network cluster_host
+    (* If cluster_host exists on master, we simply return the network of its PIF *)
+  | None ->
+    info "no cluster_host on master, let's check the cluster_hosts";
+    (* Otherwise, we try to find a common network among the cluster_host PIFs *)
+    match Db.Cluster_host.get_all ~__context with
```

Comment:
If we want to be even more general & correct, than we should get the cluster hosts of the cluster `self` - this argument wasn't and isn't currently used.

Reply:
Yes, we could have multiple clusters, we should not expect them all to be on same network. Would be useful to write a helper function that gives you all the cluster hosts for a particular cluster.

Reply:
We can use the db getter: `Db.Cluster.get_cluster_hosts`